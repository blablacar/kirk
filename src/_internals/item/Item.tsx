import React from 'react'
import cc from 'classcat'

import { color } from '../../_utils/branding'
import { A11yProps, pickA11yProps } from '../../_utils/interfaces'
import { Button } from '../../button/Button'
import { ChevronIcon } from '../../icon/chevronIcon'
import { NormalizeProps } from '../../layout/layoutNormalizer'
import { Text, TextDisplayType, TextTagType } from '../../text'

export enum ItemStatus {
  DEFAULT = 'default',
  LOADING = 'loading',
  CHECKED = 'checked',
}

export type ItemProps = A11yProps &
  NormalizeProps &
  Readonly<{
    chevron?: boolean
    className?: string
    href?: string | JSX.Element
    highlighted?: boolean
    isClickable?: boolean
    isWrappable?: boolean
    leftTitle?: React.ReactNode
    leftTitleButtonAddon?: React.ReactElement<Button>
    leftTitleDisplay?: TextDisplayType
    leftTitleColor?: string
    leftBody?: string | React.ReactNode
    leftBodyDisplay?: TextDisplayType
    leftBodyColor?: string
    leftBodyAnnotation?: React.ReactNode
    leftBodyAnnotationDisplay?: TextDisplayType
    leftBodyAnnotationColor?: string
    leftAddon?: React.ReactNode
    rightTitle?: string | JSX.Element
    rightTitleDisplay?: TextDisplayType
    rightTitleStrikeThrough?: boolean
    rightTitleAriaProps?: A11yProps
    rightTitleColor?: string
    rightBody?: React.ReactNode
    rightBodyDisplay?: TextDisplayType
    rightBodyColor?: string
    rightAddon?: React.ReactNode
    tag?: JSX.Element
    onClick?: (event: React.MouseEvent<HTMLElement>) => void
    onBlur?: (event: React.FocusEventHandler<HTMLElement>) => void
    onFocus?: (event: React.FocusEventHandler<HTMLElement>) => void
    onMouseDown?: (event: React.MouseEvent<HTMLElement>) => void
    hideHoverBackground?: boolean
    disabled?: boolean
  }>

export const Item = (props: ItemProps) => {
  const {
    chevron,
    className,
    href,
    onClick,
    onBlur,
    onFocus,
    onMouseDown,
    highlighted,
    isClickable,
    isWrappable,
    leftTitle,
    leftTitleButtonAddon,
    rightTitleStrikeThrough,
    rightTitleAriaProps,
    leftTitleDisplay = TextDisplayType.TITLE,
    leftTitleColor,
    leftBody,
    leftBodyDisplay = TextDisplayType.BODY,
    leftBodyColor,
    leftBodyAnnotation,
    leftBodyAnnotationDisplay,
    leftBodyAnnotationColor,
    leftAddon,
    rightTitle,
    rightTitleDisplay = TextDisplayType.TITLE,
    rightTitleColor,
    rightBody,
    rightBodyDisplay = TextDisplayType.BODY,
    rightBodyColor,
    rightAddon,
    tag = <div />,
    hideHoverBackground = false,
    disabled = false,
  } = props
  const a11yAttrs = pickA11yProps<ItemProps>(props)
  let Tag = tag.type
  let tagProps = tag.props

  if (href) {
    if (typeof href !== 'string') {
      Tag = href.type
      tagProps = { ...tagProps, ...href.props }
    } else {
      Tag = 'a'
      tagProps = { href }
    }
  }

  if (Tag === 'button') {
    // The code below make sure that if we use a HTML button, a proper type='button' or
    // type='submit' is added. type='button' is the default.
    // This type=button is needed to make sure non-submit buttons are not activated when
    // submitting forms.
    // See: https://stackoverflow.com/questions/4763638/enter-triggers-button-click
    const buttonType = tag.props.type || 'button'
    tagProps = { ...tagProps, type: buttonType }
  }

  const hasRightText = rightTitle || rightBody

  const getTextColor = (textColor: string) => (disabled ? color.gray : textColor)

  const getLeftTitle = (value: React.ReactNode): React.ReactNode => {
    if (React.isValidElement(value)) {
      return leftTitle
    }
    return (
      <Text
        className={leftTitleButtonAddon ? 'kirk-item-title--withButtonAddon' : null}
        display={leftTitleDisplay}
        textColor={getTextColor(leftTitleColor)}
        tag={TextTagType.SPAN}
      >
        {leftTitle}
        {/* In case of a clickable Item don't display the addon button */}
        {!href && leftTitleButtonAddon}
      </Text>
    )
  }

  return (
    <Tag
      {...tagProps}
      onClick={onClick}
      onFocus={onFocus}
      onBlur={onBlur}
      onMouseDown={onMouseDown}
      className={cc([
        'kirk-item',
        {
          'kirk-item--highlighted': highlighted,
          'kirk-item--clickable': isClickable,
          'kirk-item--wrappable': isWrappable,
          'kirk-item--hideHoverBackground': hideHoverBackground,
        },
        className,
      ])}
      {...a11yAttrs}
    >
      <span className="kirk-item-leftWrapper">
        {leftAddon && <span className="kirk-item-leftAddon">{leftAddon}</span>}
        <span className="kirk-item-leftText">
          {leftTitle && getLeftTitle(leftTitle)}
          {leftBody && (
            <Text
              className="kirk-item-body"
              display={leftBodyDisplay}
              textColor={getTextColor(leftBodyColor)}
              tag={TextTagType.SPAN}
            >
              {leftBody}
            </Text>
          )}
          {leftBodyAnnotation && (
            <Text
              className="kirk-item-body-annotation"
              display={leftBodyAnnotationDisplay}
              textColor={getTextColor(leftBodyAnnotationColor)}
              tag={TextTagType.SPAN}
            >
              {leftBodyAnnotation}
            </Text>
          )}
        </span>
      </span>
      {hasRightText && (
        <span className="kirk-item-rightText">
          {rightTitle && (
            <Text
              className={cc([
                'kirk-item-title',
                {
                  'kirk-item--strikethrough': rightTitleStrikeThrough,
                },
              ])}
              display={rightTitleDisplay}
              textColor={getTextColor(rightTitleColor)}
              tag={TextTagType.DIV}
              {...rightTitleAriaProps}
            >
              {rightTitle}
            </Text>
          )}
          {rightBody && (
            <Text
              className="kirk-item-body"
              display={rightBodyDisplay}
              textColor={getTextColor(rightBodyColor)}
              tag={TextTagType.DIV}
            >
              {rightBody}
            </Text>
          )}
        </span>
      )}
      {rightAddon && <span className="kirk-item-rightAddon">{rightAddon}</span>}
      {chevron && (
        <span className="kirk-item-rightAddon">
          <ChevronIcon iconColor={!isClickable ? color.gray : color.lightMidnightGreen} />
        </span>
      )}
    </Tag>
  )
}
