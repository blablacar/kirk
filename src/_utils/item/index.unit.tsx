import React from 'react'
import renderer from 'react-test-renderer'

import Item from './index'
import ClockIcon from 'icon/clockIcon'
import ChevronIcon from 'icon/chevronIcon'

it('Should not have changed', () => {
  const item = renderer
    .create(<Item className="custom" leftTitle="Test" leftBody="Test" rightAddon={<ClockIcon />} />)
    .toJSON()
  expect(item).toMatchSnapshot()
})

it('Should be displayed in default state without props', () => {
  const wrapper = shallow(<Item />)
  expect(wrapper.hasClass('kirk-item')).toBe(true)
  expect(wrapper.type()).toBe('div')
})

it('Should accept a custom `className`', () => {
  const customClassName = 'custom'
  const wrapper = shallow(<Item className={customClassName} />)
  expect(wrapper.hasClass(customClassName)).toBe(true)
})

it('Should be highlighted', () => {
  const wrapper = shallow(<Item highlighted />)
  expect(wrapper.hasClass('kirk-item--highlighted')).toBe(true)
})

it('Should accept a custom `tag`', () => {
  const wrapper = shallow(<Item tag={<li />} />)
  expect(wrapper.type()).toBe('li')
})

it('Should not return an href prop when no href is pass to the item', () => {
  const wrapper = shallow(<Item />)
  expect(wrapper.prop('href')).toBeUndefined()
})

it('Should return a tag A when href is a string', () => {
  const wrapper = shallow(<Item href="#" />)
  expect(wrapper.type()).toBe('a')
  expect(wrapper.prop('href')).toBe('#')
})

it('Should return a tag of a with its href when href is a component a', () => {
  const wrapper = shallow(<Item href={<a href="#" />} />)
  expect(wrapper.type()).toEqual('a')
  expect(wrapper.prop('href')).toEqual('#')
})

it('Should display a left add-on', () => {
  const wrapper = shallow(<Item leftAddon={<ClockIcon />} />)
  expect(wrapper.find(ClockIcon).exists()).toBe(true)
})

it('Should display title on the left', () => {
  const wrapper = mount(<Item leftTitle="Test" />)
  expect(wrapper.find('.kirk-item-leftText .kirk-text-title').exists()).toBe(true)
})

it('Should display body text on the left', () => {
  const wrapper = mount(<Item leftBody="Test" />)
  expect(wrapper.find('.kirk-item-leftText .kirk-text-body').exists()).toBe(true)
})

it('Should display a right add-on', () => {
  const wrapper = shallow(<Item rightAddon={<ClockIcon />} />)
  expect(wrapper.find(ClockIcon).exists()).toBe(true)
})

it('Should display title on the right', () => {
  const wrapper = mount(<Item rightTitle="Test" />)
  expect(wrapper.find('.kirk-item-rightText .kirk-text-title').exists()).toBe(true)
})

it('Should display body text on the right', () => {
  const wrapper = mount(<Item rightBody="Test" />)
  expect(wrapper.find('.kirk-item-rightText .kirk-text-body').exists()).toBe(true)
})

it('Should display a chevron', () => {
  const wrapper = shallow(<Item chevron />)
  expect(wrapper.find(ChevronIcon).exists()).toBe(true)
})
