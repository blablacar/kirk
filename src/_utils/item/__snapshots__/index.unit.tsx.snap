// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Should not have changed 1`] = `
.c0 {
  margin: 0;
  font-weight: 400;
}

.c0.kirk-text-button {
  color: #054752;
  font-size: 16px;
  line-height: 20px;
}

.c0.kirk-text-body,
.c0.kirk-text-bodyStrong {
  color: #708C91;
  font-size: 16px;
  line-height: 20px;
}

.c0.kirk-text-caption {
  color: #708C91;
  font-size: 13px;
  line-height: 16px;
}

.c0.kirk-text-title,
.c0.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c0.kirk-text-display1 {
  color: #054752;
  font-size: 82px;
  line-height: 82px;
}

.c0.kirk-text-display2 {
  color: #054752;
  font-size: 30px;
  line-height: 1.06;
}

.c0.kirk-text-subheader,
.c0.kirk-text-subheaderStrong {
  color: #054752;
  font-size: 22px;
  line-height: 24px;
}

.c0.kirk-text-title,
.c0.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c0.kirk-text-bodyStrong,
.c0.kirk-text-subheaderStrong,
.c0.kirk-text-titleStrong {
  font-weight: 500;
}

<div
  className="jsx-1542968643 kirk-item custom"
>
  <div
    className="jsx-1542968643 kirk-item-leftText"
  >
    <div
      className="kirk-text kirk-text-title kirk-item-title c0"
    >
      Test
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body c0"
    >
      Test
    </div>
  </div>
  <div
    className="jsx-1542968643 kirk-item-rightAddon"
  >
    <svg
      aria-hidden={true}
      className="kirk-icon"
      fill="#708C91"
      height={24}
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      
      <g
        fill="none"
        fillRule="evenodd"
      >
        <path
          d="M0 0h24v24H0z"
        />
        <g
          stroke="#708C91"
          strokeLinecap="round"
          strokeLinejoin="round"
        >
          <circle
            cx="12"
            cy="12"
            r="9.565"
          />
          <path
            d="M12 6.783V12h5.217"
          />
        </g>
      </g>
    </svg>
  </div>
  <style />
</div>
`;
